cEP-0019.md

```diff
@@ -0,0 +1,454 @@
+              }
+            }
+          ]
+        }
+      }
+    }
+  }
+}
+```
+
+When a new issue/pull request is detected, gh-board will fetch that issue and
+loads it asynchronously. Meta-reviews also need to be fetched during this
+procedure.
+
+### Analysis of Meta-reviews
+
+This section explains how meta-review information can be utilized to
+generate a meaningful analysis.
+
+#### Ranking list
+
+A ranking list would be displayed on coala community website, according
+to points people get. The formula for points would be as follows:
+
+P1 = total number of THUMBS_UP 👍 a person gets for all reviews he did.
+
+P2 = total number of THUMBS_UP 👍 a person gives to other people for
+their reviews.
+
+N1 = total number of THUMBS_DOWN 👎 a person gets for all reviews he did.
+
+N2 = total number of THUMBS_DOWN 👎 a person gives to other people for
+their reviews.
+
+Note P2 and N2 are used to encourage people to get involved in this
+reactions feedback system.
+
+Then final score S would be:
+
+S = k * (P1 - N1) + c1 * P2 + c2 \* N2
+
+c1 = 0.2, c2 = 0.4, are coefficients. c2 is larger than c1 because people
+are usually reluctant to give negative reactions.
+
+k is the impact factor. The higher score a person has, the more impacts he has,
+thus his meta-reviews are more valuable. k is based on the scores in the
+previous iteration and is normalized. For example, in a previous iteration,
+Alice got 2 marks, Bob got 0.8 marks and Charlie got 10 marks. The calculation
+procedure would be as follows:
+
+```python
+>>> c = [2, 0.8, 10]
+>>> max_score = max(c)
+>>> min_score = min(c)
+>>> result = [(i - min_score) / (max_score - min_score) for i in c]
+>>> print(result)
+[0.13043478260869565, 0.0, 1.0]
+>>> result_adjust = [i + 0.1 for i in result] # add bias
+>>> print(result_adjust)
+[0.23043478260869565, 0.1, 1.1]
+>>> result_rounded = [round(i, 3) for i in result_adjust]
+>>> print(result_rounded)
+[0.23, 0.1, 1.1]
+```
+
+Therefore, the k factor for Alice, Bob and Charlie in the next iteration would
+be: 0.23, 0.1 and 1.1, respectively.
+
+If a person has zeros for all factors, then he is regarded as an inactive
+reviewer and would not be displayed. If a person has negative score, he will
+still be displayed, but the meta-reviews he gives to other people will not
+count.
+
+#### Meta-reviews in Need
+
+##### Definition
+
+When is a review needing meta-review? There are two perspectives:
+
+1. Meta-review by developers, i.e. all people within the organization. The
+   'developer' concept here is different from that used in coala. Instead,
+   anyone with write access (including maintainers) to the repo should be
+   regarded as developers. For developers, review comments before merge/close
+   need meta-reviews. Those review comments are sorted according to update
+   time. In particular, one will always see the review comment that is under
+   his PR first.
+
+2. Meta-review by maintainers. For maintainers, review comments after
```

Comment:
Another important technical detail of meta-review in coala on GitHub is that the context disappears on an amend push, so meta-review by maintainers after merge may be without the necessary context to be properly evaluated.  There is one way to avoid that is for a bot (or human) to mention the commit SHA in a comment, which means GitHub will keep it in its list of dangling commits.  This commit SHA mentioning could be done in the 'review summary', and a bot might even be able to amend the humans review summaries to append the commit it was referring to.

This might be a gitmate plugin which is needed to make your project functional.  Or maybe there are other existing tools which do this.  Or maybe you build a prototype in [probot](https://github.com/probot/probot), and re-implementation in gitmate is left until after the project.

Reply:
I'm not familiar enough with GitHub, but doesn't the commit disappear on GitHub after an amend push?

Reply:
Yes, the commit disappears on GitHub, completely, if it hasnt been mentioned.  fwiw, GitLab keeps them.
If a bot mentions the commit sha, GitHub keeps the commit even if it is no longer in the PR.
So we may need a bot which mentions the sha when the commits are pushed, or when first reviewed, so that latter force pushes dont delete the content.

mentioning every commit when pushed is not ideal, as it is common for people to notice an error quickly and fix it themselves before any reviews.  Keeping a history of them all may encourage people to be more careful before pushing, but likely will make them a bit more scared of pushing, and reluctant to do it.